File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <utility>
#include <cmath>
using namespace std;

// O(nlogE)

int dijkstra(int n, vector<vector<int>> &adj){
    // 1번 노드로부터 최단 경로
    vector<int> dist(n+1, 20001);
    // distance, node
    priority_queue<pair<int,int>> pq;
    pq.push({0, 1});
    dist[1] = 0;
    
    while(!pq.empty()){
        int cur_dist = pq.top().first;
        int cur_node = pq.top().second;
        pq.pop();
        
        // 최단 거리 아니면 스킵
        if(cur_dist > dist[cur_node]) continue;
        
        for(int i=0; i<adj[cur_node].size(); i++){
            int next_node = adj[cur_node][i];
            if(dist[next_node] > cur_dist + 1){
                dist[next_node] = cur_dist+1;
                pq.push({cur_dist+1, next_node});
            }
        }        
    }
    
    // 최대 길이 체크 
    int max_size = 0;
    for(int i=1; i<=n; i++){
        max_size = max(max_size, dist[i]);
    }
    
    int ans = 0;
    for(int i=1; i<=n; i++){
        if(dist[i] == max_size){
            ans++;
        }
    }
    return ans;
}


int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    // 인접 여부 판단
    vector<vector<int>> adj(n+1);
    for(int i=0; i<edge.size(); i++){
        int e = edge[i][0];
        int v = edge[i][1];
        adj[e].push_back(v);
        adj[v].push_back(e);
    }
    answer = dijkstra(n, adj);
    return answer;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;


// 1번 노드로부터 가장 멀리 떨어진 
// 1번과 이어진 -> 이후 노드로부터 이어진 -> 처음 값이 갱신되면 최단거리 

class Node implements Comparable<Node>{
    int dist = 0;
    int node;
    public Node(int dist, int node){
        this.dist = dist;
        this.node = node;
    }
    
    @Override
    public int compareTo(Node b){
        return Integer.compare(this.dist, b.dist);
    }
}

class Solution {
    int max_val = 30000;
    public int solution(int n, int[][] edge) {
        
        ArrayList<Integer>[] adj = new ArrayList[n+1];
        for(int i=0; i<adj.length; i++){
            adj[i] = new ArrayList<Integer>();
        }
        
        for(int i=0; i<edge.length; i++){
            int n1 = edge[i][0];
            int n2 = edge[i][1];
            adj[n1].add(n2);
            adj[n2].add(n1);
        }
        
        
        int[] weights = new int[n+1];
        for(int i=1; i<=n; i++){
            weights[i] = max_val;
        }
        weights[1] = 0;
        // dist, node
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(0,1));
        
        while(!pq.isEmpty()){
            Node cur = pq.poll();
            int cur_dist = cur.dist;
            int cur_node = cur.node;
            if(cur_dist > weights[cur_node]) continue;
            
            for(int next_node: adj[cur_node]){
                if(weights[next_node] > cur_dist + 1){
                    weights[next_node] = cur_dist + 1;
                    pq.add(new Node(cur_dist+1, next_node));
                }
            }
        }
        
        int most_val = -1;
        int answer = 0;
        for(int i=1; i<=n; i++){
            if(most_val < weights[i]){
                most_val = weights[i];
                answer = 0;
            }
            if(most_val == weights[i]){
                answer++;
            }
        }
        return answer;
    }
}